lib/FreeRTOS/mqtt-agent/mqtt_agent.c

```diff
@@ -457,145 +464,81 @@ static MQTTStatus_t addCommandToQueue( AgentMessageContext_t * pQueue,
-                addAckToList = ( pPublishInfo->qos != MQTTQoS0 ) && ( operationStatus == MQTTSuccess );
-
-                break;
-
-            case SUBSCRIBE:
-            case UNSUBSCRIBE:
-                pSubscribeArgs = ( MQTTAgentSubscribeArgs_t * ) ( pCommand->pArgs );
-                packetId = MQTT_GetPacketId( pMQTTContext );
-
-                if( pCommand->commandType == SUBSCRIBE )
-                {
-                    /* Even if some subscriptions already exist in the subscription list,
-                     * it is fine to send another subscription request. A valid use case
-                     * for this is changing the maximum QoS of the subscription. */
-                    operationStatus = MQTT_Subscribe( pMQTTContext,
-                                                      pSubscribeArgs->pSubscribeInfo,
-                                                      pSubscribeArgs->numSubscriptions,
-                                                      packetId );
-                }
-                else
-                {
-                    operationStatus = MQTT_Unsubscribe( pMQTTContext,
-                                                        pSubscribeArgs->pSubscribeInfo,
-                                                        pSubscribeArgs->numSubscriptions,
-                                                        packetId );
-                }
-
-                addAckToList = ( operationStatus == MQTTSuccess );
-                break;
-
-            case PING:
-                operationStatus = MQTT_Ping( pMQTTContext );
-
-                break;
-
-            case CONNECT:
-                pConnectArgs = ( MQTTAgentConnectArgs_t * ) ( pCommand->pArgs );
-                operationStatus = MQTT_Connect( pMQTTContext,
-                                                pConnectArgs->pConnectInfo,
-                                                pConnectArgs->pWillInfo,
-                                                pConnectArgs->timeoutMs,
-                                                &( pConnectArgs->sessionPresent ) );
-                break;
-
-            case DISCONNECT:
-                operationStatus = MQTT_Disconnect( pMQTTContext );
-                runProcessLoops = false;
-
-                break;
+        assert( pCommand->commandType < NUM_COMMANDS );
+        commandFunction = pCommandFunctionTable[ pCommand->commandType ];
+        pCommandArgs = pCommand->pArgs;
+    }
+    else
+    {
+        commandFunction = pCommandFunctionTable[ NONE ];
+    }
 
-            case TERMINATE:
-                LogInfo( ( "Terminating command loop.\n" ) );
-                runProcessLoops = false;
+    operationStatus = commandFunction( pMqttAgentContext, pCommandArgs, &commandOutParams );
 
-            default:
-                break;
-        }
+    if( commandOutParams.addAcknowledgment )
+    {
+        ackAdded = addAwaitingOperation( pMqttAgentContext, commandOutParams.packetId, pCommand );
```

Comment:
Can this also be moved to the proxy functions?

Reply:
Adding yes, removing unsure. The event callback is responsible for removing the acks, and I'm not sure if that should be moved here.

Reply:
The reason for the question was to see if the command processing logic can be completely independent of the MQTT. I was expecting the implementation to be complete with line 493 and returning the appropriate status. I understand the issues that you mentioned about the inconsistency in adding and removing acks will be in different files. I think if we are not planning to separate out command processing with MQTT, the current logic should be ok.  I will leave the comment open so that we will get another look by a PRBR on this topic.